xls/modules/dbe/lz4_encoder.x

```diff
@@ -0,0 +1,494 @@
+                FIFO_IN_SZ: u32>()
+                -> State<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH, HASH_BITS,
+                            MINMATCH, FINAL_LITERALS, FIFO_CACHE_SZ,
+                            FIFO_IN_SZ> {
+    State{
+        fifo_cache: uN[SYM_WIDTH][FIFO_CACHE_SZ]: [uN[SYM_WIDTH]:0, ...],
+        fifo_in: uN[SYM_WIDTH][FIFO_IN_SZ]: [uN[SYM_WIDTH]:0, ...],
+        fifo_in_nvalid: u32:0,
+        wp: uN[PTR_WIDTH]:0,
+        hb_all_valid: false,
+        pnew: uN[PTR_WIDTH]:0,
+        pold: uN[PTR_WIDTH]:0,
+        ht_ptr: uN[HASH_BITS]:0,
+        recycle: false,
+        eof: false,
+        fsm: FsmSt::RESET
+    }
+}
+
+pub proc encoder<SYM_WIDTH: u32, PTR_WIDTH: u32, CNT_WIDTH: u32, HASH_BITS: u32,
+                    MINMATCH: u32 = {u32:4}, FINAL_LITERALS: u32 = {u32:12},
+                    FIFO_CACHE_SZ: u32 = {MINMATCH - u32:1},
+                    FIFO_IN_SZ: u32 = {std::umax(MINMATCH, FINAL_LITERALS)},
+                    HB_RAM_SZ: u32 = {u32:1 << PTR_WIDTH},
+                    HT_RAM_SZ: u32 = {u32:1 << HASH_BITS}> {
+    i_data: chan<PlainData<SYM_WIDTH>> in;
+    o_encoded: chan<Token<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>> out;
+
+    /// History buffer RAM
+    /// Size: (1<<PTR_WIDTH) x SYM_WIDTH
+    o_ram_hb_req: chan<RamReq<PTR_WIDTH, SYM_WIDTH, 1>> out;
+    i_ram_hb_resp: chan<RamResp<SYM_WIDTH>> in;
+
+    /// Hash table RAM
+    /// Size: (1<<HASH_BITS) x PTR_WIDTH
+    o_ram_ht_req: chan<RamReq<HASH_BITS, PTR_WIDTH, 1>> out;
+    i_ram_ht_resp: chan<RamResp<PTR_WIDTH>> in;
+
+    init {
+        init_state<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH, HASH_BITS,
+                    MINMATCH, FINAL_LITERALS, FIFO_CACHE_SZ, FIFO_IN_SZ>()
+    }
+
+    config (
+        // Data in, tokens out
+        i_data: chan<PlainData<SYM_WIDTH>> in,
+        o_encoded: chan<Token<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>> out,
+        // RAM
+        o_ram_hb_req: chan<RamReq<PTR_WIDTH, SYM_WIDTH, 1>> out,
+        i_ram_hb_resp: chan<RamResp<SYM_WIDTH>> in,
+        o_ram_ht_req: chan<RamReq<HASH_BITS, PTR_WIDTH, 1>> out,
+        i_ram_ht_resp: chan<RamResp<PTR_WIDTH>> in,
+    ) {
+        // let (o_ram_hb_req, i_ram_hb_req) = chan<RamReq<PTR_WIDTH, SYM_WIDTH, u32:1>>;
+        // let (o_ram_hb_resp, i_ram_hb_resp) = chan<RamResp<SYM_WIDTH>>;
+        // spawn ram::RamModel<SYM_WIDTH, HB_RAM_SZ, SYM_WIDTH>(
+        //     i_hb_read_req, o_hb_read_resp, i_hb_write_req, o_hb_write_resp);
+
+        // let (o_ram_ht_req, i_ram_ht_req) = chan<RamReq<HASH_BITS, PTR_WIDTH, u32:1>>;
+        // let (o_ram_ht_resp, i_ram_ht_resp) = chan<RamResp<PTR_WIDTH>>;
+        // spawn ram::RamModel<PTR_WIDTH, HT_RAM_SZ, PTR_WIDTH>(
+        //     i_ht_read_req, o_ht_read_resp, i_ht_write_req, o_ht_write_resp);
+
+        (
+            i_data, o_encoded,
+            o_ram_hb_req, i_ram_hb_resp,
+            o_ram_ht_req, i_ram_ht_resp,
+        )
+    }
+
+    next (tok: token, cur: State) {
```

Comment:
@proppy 

I think this is a good example of why #1037 is needed.

What I'm trying to do here:
- In some FSM states (e.g. START_MATCH_0), HB RAM is written (line 194), and then HT RAM is read (line 234).
- In other FSM states (e.g. START_MATCH_1), HT RAM is written (line 245), and then HB RAM is read (line 262).

It's easy to prove that above states are mutually exclusive. However, I still can not use two mutually exclusive send's or recv's to the same channel. Replacing them with `send_if` / `recv_if` manually is not enough: the current XLS limitation seems to be that I must provide a _single_ `send_if` or `recv_if` for each channel. There must not be two mutually exclusive `send_if`'s, there must be at most one `send_if` in the whole `next()` function for each channel. While it is possible to rewrite the code to fulfill that constraint, the code most probably will get quite cumbersome and dirty as a result.. 

I see another option: instead of relying on 1RW memory, I should be able to use 1R1W memories here, and because those have separate channels for read and write requests, I may be able to figure out something that makes XLS happy without much code reshuffling.

Do you know what's the status of 1R1W RAM Verilog codegen support? From XLS soruces it seems that it may be supported, but I haven't found any 1R1W DSLX example that actually generates Verilog.